tests/test_slurm_system.py

```diff
@@ -142,3 +143,88 @@ def test_update_node_states_with_mocked_outputs(mock_get_sinfo, mock_get_squeue,
 def test_parse_node_list(node_list: str, expected_parsed_node_list: List[str]):
     parsed_node_list = parse_node_list(node_list)
     assert parsed_node_list == expected_parsed_node_list
+
+
+def setup_mock_slurm_system(slurm_system, group_name):
+    """
+    Helper function to set up a mock Slurm system with nodes and their states.
+    Args:
+        slurm_system: The Slurm system instance.
+        group_name: Name of the group to set up within the partition.
+    """
+    partition_name = "backup"
+
+    mock_nodes = [
+        SlurmNode(name="node01", partition=partition_name, state=SlurmNodeState.IDLE),
+        SlurmNode(name="node02", partition=partition_name, state=SlurmNodeState.COMPLETING),
+        SlurmNode(name="node03", partition=partition_name, state=SlurmNodeState.IDLE),
+        SlurmNode(name="node04", partition=partition_name, state=SlurmNodeState.DOWN),
+    ]
+
+    slurm_system.partitions = [
+        SlurmPartition(
+            name=partition_name,
+            nodes=[node.name for node in mock_nodes],
+            groups=[SlurmGroup(name=group_name, nodes=[node.name for node in mock_nodes])],
+        )
+    ]
+
+    slurm_system.partitions[0]._slurm_nodes = mock_nodes
+
+    grouped_nodes = {
+        SlurmNodeState.IDLE: [
+            slurm_system.groups[partition_name][group_name][0],
+            slurm_system.groups[partition_name][group_name][2],
+        ],
+        SlurmNodeState.COMPLETING: [
+            slurm_system.groups[partition_name][group_name][1],
+        ],
+        SlurmNodeState.ALLOCATED: [],
+    }
+
+    return grouped_nodes
+
+
+def test_allocate_nodes_max_avail(slurm_system):
+    group_name = "backup"
```

Comment:
Variable name is misleading, isn't it a partition name? (same for another test)

Reply:
From the function allocate_nodes : 
"""
        Allocate nodes based on the requested number or maximum availability.

        Args:
            grouped_nodes (Dict[SlurmNodeState, List[SlurmNode]]): Nodes grouped by their state.
            number_of_nodes (Union[int, str]): The number of nodes to allocate, or 'max_avail' to allocate
                all available nodes.
            **group_name (str): The name of the group.**

        Returns:
            List[SlurmNode]: A list of allocated nodes.

        Raises:
            ValueError: If the requested number of nodes exceeds the available nodes.
        """

Reply:
Ah, I see. I was confused by `groups["backup"][group_name]`. Both partition and group names are the same.

BTW, `allocate_nodes` needs `group_name` only to raise an error, which feels a bit weird, WDYT?

Reply:
Exactly I can see that almost all the functions in the slurm_system.py file have this param. For `allocate_nodes`, we only need this param so that the user knows in which group we haven't managed to allocate the number of nodes he requested. I don't think we can remove that. What do you think ?

Reply:
It is a signal that our class-internal logic is broken. Exception to be raised close to public API where we have all the inputs. If we constantly pass a variable as function arg, it should be a class member.

I don't have any good idea for reworking it right now.  And while this PR highlighted this issue, it is not a blocker here. So unless you have a good idea how to solve it, let's leave it as is.

P.S. other comments still make sense. And if we stay with `groups["backup"][group_name]` let's at lease give then better names, like `groups["partition_name"]["group_name"]` or something like that.

Reply:
I don't have a better idea right now to fix it...